src/bos/server/controllers/v2/components.py

```diff
@@ -88,47 +92,80 @@ def get_v2_components_data(id_list=None,
+                           delete_timestamp: bool = False):
     """Used by the GET /components API operation
 
     Allows filtering using a comma separated list of ids.
     """
-    response = []
-    if id_list:
-        for component_id in id_list:
-            data = DB.get(component_id)
-            if data:
-                response.append(data)
+    if any([id_list, enabled, session, staged_session, phase, status, tenant]):
+        tenant_components = None if not tenant else get_tenant_component_set(
+            tenant)
+        _component_filter_func = partial(_filter_component,
+                                         id_list=id_list or None,
+                                         enabled=enabled,
+                                         session=session or None,
+                                         staged_session=staged_session or None,
+                                         phase=phase or None,
+                                         status=status or None,
+                                         tenant_components=tenant_components
+                                         or None)
+    elif delete_timestamp:
+        _component_filter_func = partial(_set_status,
+                                         delete_timestamp=delete_timestamp)
     else:
-        # TODO: On large scale systems, this response may be too large
-        # and require paging to be implemented
-        response = DB.get_all()
-    # The status must be set before using _matches_filter as the status is one of the
-    # matching criteria.
-    response = [_set_status(r) for r in response if r]
-    if enabled is not None or session is not None or staged_session is not None or \
-       phase is not None or status is not None:
-        response = [
-            r for r in response if _matches_filter(
-                r, enabled, session, staged_session, phase, status)
-        ]
-    if tenant:
-        tenant_components = get_tenant_component_set(tenant)
-        limited_response = [
-            component for component in response
-            if component["id"] in tenant_components
-        ]
-        response = limited_response
-    return response
+        _component_filter_func = _set_status
+
+    return DB.get_all_filtered(filter_func=_component_filter_func,
+                               start_after_key=start_after_id,
+                               page_size=page_size)
 
 
-def _set_status(data):
+def _filter_component(data: dict,
+                      id_list=None,
+                      enabled=None,
+                      session=None,
+                      staged_session=None,
+                      phase=None,
+                      status=None,
+                      tenant_components=None,
+                      delete_timestamp: bool = False) -> dict | None:
+    # Do all of the checks we can before calculating status, to avoid doing it needlessly
+    if id_list is not None and data["id"] not in id_list:
```

Comment:
if id_list is not None is same as if id_list
Similarly other lines can be updated.

Reply:
This is true, and probably not a big deal, but I believe the check that it is not None is faster than the generic boolean check. Part of what I was trying to do here was to make this filter function as efficient as possible, because it is potentially being called thousands of times in any given request.